lazy-pages/src/sys.rs

```diff
@@ -39,117 +53,284 @@ cfg_if! {
 pub struct ExceptionInfo {
     /// Address where fault is occurred
     pub fault_addr: *const (),
+    pub is_write: Option<bool>,
 }
 
-/// Returns key which `page` has in storage.
-/// `prefix` is current program prefix in storage.
-fn page_key_in_storage(prefix: &Vec<u8>, page: PageNumber) -> Vec<u8> {
-    let mut key = Vec::with_capacity(prefix.len() + std::mem::size_of::<u32>());
-    key.extend(prefix);
-    key.extend(page.0.to_le_bytes());
-    key
+/// Struct for fast calculation of page key in storage.
+/// Key consists of two parts:
+/// 1) current program prefix in storage
+/// 2) page number in little endian bytes order
+/// First part is always the same, so we can copy it to buffer
+/// once and then use it for all pages.
+struct PagePrefix {
+    buffer: Vec<u8>,
+}
+
+impl PagePrefix {
+    /// New page prefix from program prefix
+    pub fn new_from_program_prefix(program_prefix: &[u8]) -> Self {
+        let mut prefix = Self {
+            buffer: Vec::with_capacity(program_prefix.len() + std::mem::size_of::<u32>()),
+        };
+        prefix.buffer.extend(program_prefix);
+        prefix.buffer.extend(u32::MAX.to_le_bytes());
+        prefix
+    }
+    /// Returns key in storage for `page`.
+    pub fn calc_key_for_page(&mut self, page: PageNumber) -> &[u8] {
+        let len = self.buffer.len();
+        self.buffer[len - std::mem::size_of::<u32>()..len]
+            .copy_from_slice(page.0.to_le_bytes().as_slice());
+        &self.buffer
+    }
+}
+
+/// Wasm address wrapper, for which we can be sure
+/// it's in safe/checked state, so we can avoid some
+/// checks while using this addr.
+struct CheckedWasmAddr {
+    wasm_mem_addr: usize,
+    wasm_mem_size: u32,
+    addr: WasmAddr,
+}
+
+impl CheckedWasmAddr {
+    pub fn new_from_native(
+        native_addr: usize,
+        wasm_mem_addr: usize,
+        stack_end: WasmAddr,
+        wasm_mem_size: u32,
+    ) -> Result<Self, Error> {
+        let calc_wasm_mem_end_addr = || {
```

Comment:
I think it's simpler to calculate address right here

Reply:
Done